File: python/sense/goal_to_sense.py
"""GoalSpec → additive SensoryInjection. Mirrors js/sense/goalToSense.js."""

from __future__ import annotations

import json
from pathlib import Path


def load_goal(path: str | Path) -> dict:
    return json.loads(Path(path).read_text())


def _clone(frame: dict) -> dict:
    out = dict(frame)
    out["luminance"] = list(frame.get("luminance") or [])
    out["objects"] = [dict(o) for o in frame.get("objects") or []]
    out["odor"] = dict(frame.get("odor") or {})
    out["taste"] = dict(frame.get("taste") or {})
    out["objectChannels"] = dict(frame.get("objectChannels") or {})
    return out


def _nearest(world: dict | None, query: dict | None):
    if not world or not query:
        return None
    name = query.get("name")
    kind = query.get("kind")
    pos = world.get("position") or {"x": 0, "y": 0, "z": 0}
    if kind == "entity":
        pool = world.get("entities") or []
    elif kind == "item":
        pool = world.get("items") or []
    else:
        pool = world.get("blocks") or []
    best = None
    best_d = 1e18
    for src in pool:
        n = str(src.get("name") or src.get("type") or "").replace("minecraft:", "")
        if n != name and not (kind == "entity" and src.get("type") == name):
            continue
        d = ((src["x"] - pos["x"]) ** 2 + (src["y"] - pos["y"]) ** 2 + (src["z"] - pos["z"]) ** 2) ** 0.5
        if d < best_d:
            best_d = d
            best = {**src, "distance": d, "name": n}
    return best


def _weight(inj: dict, world: dict | None) -> float:
    fall = inj.get("falloff") or "tonic"
    strength = float(inj.get("strength") or 0)
    if fall == "tonic":
        return strength
    hit = _nearest(world, inj.get("source_query"))
    if not hit:
        return 0.0
    if fall == "when_seen":
        return strength
    if fall == "when_near":
        return strength / (1.0 + hit["distance"])
    if fall == "distance":
        import math
        return strength * math.exp(-hit["distance"] / float(inj.get("lambda") or 6))
    return strength


def apply_goal(frame: dict, goal: dict | None, world: dict | None = None) -> dict:
    out = _clone(frame)
    if not goal:
        return out
    for inj in goal.get("injections") or []:
        w = _weight(inj, world)
        if w <= 0:
            continue
        mod = inj.get("modality")
        if mod == "olfaction" and inj.get("pattern"):
            for glom, aff in inj["pattern"].items():
                out["odor"][glom] = out["odor"].get(glom, 0.0) + float(aff) * w
        elif mod == "gustation" and inj.get("pattern"):
            for grn, aff in inj["pattern"].items():
                out["taste"][grn] = max(out["taste"].get(grn, 0.0), float(aff) * w)
        elif mod == "vision":
            pop = inj.get("population") or "LC11"
            out["objectChannels"][pop] = min(1.5, out["objectChannels"].get(pop, 0.0) + w)
        elif mod == "mechano":
            field = inj.get("field") or "groomDust"
            out[field] = min(1.0, float(out.get(field) or 0) + w)
    return out


def goal_success(goal: dict, facts: dict) -> bool:
    if not goal or not goal.get("success"):
        return False
    s = goal["success"]
    if s["type"] == "inventory_contains":
        return (facts.get("inventory") or {}).get(s["item"], 0) >= s.get("count", 1)
    if s["type"] == "taste_contact":
        return (facts.get("taste") or {}).get(s["grn"], 0) >= s.get("min", 0.5)
    if s["type"] == "distance_above":
        d = (facts.get("distances") or {}).get(s["entity"])
        return d is None or d >= s["blocks"]
    return False


def reward_for(goal: dict, prev: dict | None, facts: dict) -> float:
    r = 0.0
    if goal_success(goal, facts) and not goal_success(goal, prev or {}):
        r += 10.0
    s = (goal or {}).get("success") or {}
    if s.get("type") == "inventory_contains":
        a = (facts.get("inventory") or {}).get(s["item"], 0)
        b = ((prev or {}).get("inventory") or {}).get(s["item"], 0)
        r += (a - b) * 5.0
    if facts.get("closerToQuery") is not None:
        r += float(facts["closerToQuery"])
    return r

File: js/sense/goalToSense.js
const { addOdor, addTaste } = require('./frame')
const { stripNs } = require('./senseBridge')

function cloneFrame (frame) {
  return {
    ...frame,
    luminance: frame.luminance.slice(),
    objects: frame.objects.map(o => ({ ...o })),
    odor: { ...frame.odor },
    taste: { ...frame.taste },
    objectChannels: { ...(frame.objectChannels || {}) },
  }
}

function nearestMatch (world, query) {
  if (!query || !world) return null
  const name = query.name
  const kind = query.kind
  const pos = world.position || { x: 0, y: 64, z: 0 }
  let best = null
  let bestD = Infinity
  const list = kind === 'entity' ? (world.entities || []) : kind === 'item' ? (world.items || []) : (world.blocks || [])
  for (const src of list) {
    const n = stripNs(src.name || src.type)
    if (n !== name && !(kind === 'entity' && (src.type === name))) continue
    const d = Math.hypot((src.x - pos.x), (src.y - pos.y), (src.z - pos.z))
    if (d < bestD) {
      bestD = d
      best = { ...src, distance: d, name: n }
    }
  }
  return best
}

function injectionWeight (inj, world) {
  const fall = inj.falloff || 'tonic'
  if (fall === 'tonic') return inj.strength
  const hit = nearestMatch(world, inj.source_query)
  if (!hit) return 0
  if (fall === 'when_seen') return inj.strength
  if (fall === 'when_near') {
    const d = hit.distance
    return inj.strength * (1 / (1 + d))
  }
  if (fall === 'distance') {
    return inj.strength * Math.exp(-hit.distance / (inj.lambda || 6))
  }
  return inj.strength
}

/**
 * Compile a GoalSpec into additive drive on a world-built SensoryFrame.
 * Does not replace world senses — only adds.
 */
function applyGoal (frame, goal, world = null) {
  const out = cloneFrame(frame)
  if (!goal || !goal.injections) return out
  for (const inj of goal.injections) {
    const w = injectionWeight(inj, world)
    if (w <= 0) continue
    if (inj.modality === 'olfaction' && inj.pattern) {
      for (const [glom, aff] of Object.entries(inj.pattern)) addOdor(out, glom, aff * w)
    } else if (inj.modality === 'gustation' && inj.pattern) {
      for (const [grn, aff] of Object.entries(inj.pattern)) addTaste(out, grn, aff * w)
    } else if (inj.modality === 'vision') {
      const pop = inj.population || 'LC11'
      out.objectChannels[pop] = Math.min(1.5, (out.objectChannels[pop] || 0) + w)
    } else if (inj.modality === 'mechano') {
      const field = inj.field || 'groomDust'
      out[field] = Math.min(1, (out[field] || 0) + w)
    }
  }
  return out
}

function goalSuccess (goal, facts) {
  if (!goal || !goal.success) return false
  const s = goal.success
  if (s.type === 'inventory_contains') {
    const n = (facts.inventory || {})[s.item] || 0
    return n >= (s.count || 1)
  }
  if (s.type === 'taste_contact') {
    return (facts.taste || {})[s.grn] >= (s.min || 0.5)
  }
  if (s.type === 'distance_above') {
    const d = facts.distances && facts.distances[s.entity]
    return d == null || d >= s.blocks
  }
  return false
}

function rewardFor (goal, prevFacts, facts) {
  if (!goal || !goal.success) return 0
  const s = goal.success
  let r = 0
  if (goalSuccess(goal, facts) && !goalSuccess(goal, prevFacts || {})) r += 10
  if (s.type === 'inventory_contains') {
    const a = (facts.inventory || {})[s.item] || 0
    const b = (prevFacts && prevFacts.inventory ? prevFacts.inventory[s.item] : 0) || 0
    r += (a - b) * 5
  }
  if (s.type === 'distance_above' && facts.distances && prevFacts && prevFacts.distances) {
    const d0 = prevFacts.distances[s.entity]
    const d1 = facts.distances[s.entity]
    if (d0 != null && d1 != null) r += (d1 - d0) * 0.1
  }
  if (facts.closerToQuery != null) r += facts.closerToQuery
  return r
}

module.exports = { applyGoal, cloneFrame, nearestMatch, injectionWeight, goalSuccess, rewardFor }
